File: 2015/Day12/main.py
import json

def sum_numbers_in_json(data):
    if isinstance(data, dict):
        return sum(sum_numbers_in_json(value) for value in data.values())
    elif isinstance(data, list):
        return sum(sum_numbers_in_json(value) for value in data)
    elif isinstance(data, int):
        return data
    return 0

def sum_numbers_in_json_without_red(data):
    if isinstance(data, dict):
        if "red" in data.values():
            return 0
        return sum(sum_numbers_in_json_without_red(value) for value in data.values())
    elif isinstance(data, list):
        return sum(sum_numbers_in_json_without_red(value) for value in data)
    elif isinstance(data, int):
        return data
    return 0

with open("input.txt", 'r') as file:
    json_data = json.load(file)


print(f"Part 1: {sum_numbers_in_json(json_data)}")
print(f"Part 2: {sum_numbers_in_json_without_red(json_data)}")

File: 2015/Day12/main.dart
import 'dart:convert';
import 'dart:io';

int sumNumbersInJson(dynamic data) {
  if (data is Map)
    return data.values.map(sumNumbersInJson).fold(0, (a, b) => a + b);
  else if (data is List)
    return data.map(sumNumbersInJson).fold(0, (a, b) => a + b);
  else if (data is int) return data;
  return 0;
}

int sumNumbersInJsonWithoutRed(dynamic data) {
  if (data is Map) {
    if (data.values.contains('red')) return 0;
    return data.values.map(sumNumbersInJsonWithoutRed).fold(0, (a, b) => a + b);
  } else if (data is List)
    return data.map(sumNumbersInJsonWithoutRed).fold(0, (a, b) => a + b);
  else if (data is int) return data;
  return 0;
}

void main() async {
  final data = json.decode(await File('input.txt').readAsString());

  print('Part 1: ${sumNumbersInJson(data)}');
  print('Part 2: ${sumNumbersInJsonWithoutRed(data)}');
}
